core/trino-main/src/main/java/io/trino/sql/gen/columnar/DictionaryAwareColumnarFilter.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.gen.columnar;
+
+import io.trino.operator.project.InputChannels;
+import io.trino.spi.Page;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.DictionaryBlock;
+import io.trino.spi.block.RunLengthEncodedBlock;
+
+import static com.google.common.base.Verify.verify;
+import static java.lang.System.arraycopy;
+
+public final class DictionaryAwareColumnarFilter
+        implements ColumnarFilter
+{
+    private final ColumnarFilter columnarFilter;
+
+    private Block lastInputDictionary;
+    private boolean[] lastOutputDictionary;
+
+    public DictionaryAwareColumnarFilter(ColumnarFilter columnarFilter)
+    {
+        verify(columnarFilter.getInputChannels().size() == 1, "Dictionary aware filtering must have only one input");
+        this.columnarFilter = columnarFilter;
+    }
+
+    @Override
+    public int filterPositionsRange(int[] outputPositions, int offset, int size, Page loadedPage)
+    {
+        Block block = loadedPage.getBlock(0);
+        if (block instanceof RunLengthEncodedBlock runLengthEncodedBlock) {
+            return processRle(outputPositions, offset, size, runLengthEncodedBlock);
+        }
+        else if (block instanceof DictionaryBlock dictionaryBlock) {
+            try {
+                return processDictionary(outputPositions, offset, size, dictionaryBlock);
+            }
+            catch (Exception ignored) {
+                // Processing of dictionary failed, but we ignore the exception here
+                // and force reprocessing of the whole block using the normal code.
+                // The second pass may not fail due to filtering.
+                // todo dictionary processing should be able to tolerate failures of unused elements
```

Comment:
I don't think this is true for columnar processing like this.  I'd drop the todo here and below